Pull request for the gene package

mnsmar opened the pull request:
Contains code for genes, transcripts, exons, introns, 5'UTR, CDS and 3'UTR.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,125 @@
+// Copyright ©2012 The bíogo Authors. All rights reserved.
```
Date.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,125 @@
+func (s *Gene) Name() string { return s.Id }
+
+// Description returns a description for the gene.
+func (s *Gene) Description() string { return "gene feature" }
```
Add a desc field and return that. The name and desc returns are the equivalent to the two parts of a FASTA desc line and should in the general case be user settable.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,125 @@
+}
+
+// Start returns the gene start on the chromosome.
+func (s *Gene) Start() int { return s.ChromStart }
```
Rename the receiver to g for *Gene methods.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,125 @@
+	ChromStart       int
+	ChromEnd         int
+	ChromOrientation feat.Orientation
+	transcripts      []Transcripter
```
I would use a []feat.Features here. Are the elements intended to be ordered?

mnsmar replied:
You mean replace field "transcripts" with "features" and change type to []feat.Features right? No I don't think they should be ordered.

kortschak replied:
Yes. Rename to features (or public equivalent that doesn't collide with the interface).

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,125 @@
+// Location returns the location of the gene. Namely the chromosome.
+func (s *Gene) Location() feat.Feature { return s.Chrom }
+
+// Orientation returns the orientation of the gene.
```
Add that this orientation is relative to the chromosome.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,125 @@
+	"errors"
+)
+
+// Assert that interfaces are satisfied
```
Add full stop.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,125 @@
+	Transcripts() []Transcripter
+}
+
+// A Gene occupies a specific area on the genome and may have 0 or more
```
s/area/region/g

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,125 @@
+func (s *Gene) Orientation() feat.Orientation { return s.ChromOrientation }
+
+// Transcripts returns a slice with all transcripts added to the gene.
+func (s *Gene) Transcripts() []Transcripter { return s.transcripts }
```
Iterate over the features and only add those that are Transcripters.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,125 @@
+
+// Features returns all features added to the gene.
+// Currently only transcripts are added so only transcripts are returned.
+func (s *Gene) Features() []feat.Feature {
```
Just return the features slice.

mnsmar replied:
The function is required by the Gener interface. 
Should I make the field public? Don't you think that we might want to have a more fine grained control (error checking) through the Add function?

kortschak replied:
The error handling is not a strong motivator. I can see a reason for satisfying feat.Set though if you have a func gene.Transcripts(feat.Set) []gene.Transcript that iterates over the features, identifying which are Transcripts and adding them to the []Transcript. Then the Transcripts method goes away.

mnsmar replied:
The `Transcripts()` method is required by the `Gener` interface. Are you suggesting to remove it from the interface? Also since I'm not making `features` a public field, this means I'm keeping the `Add` method right?

kortschak replied:
I would remove it from the interface. I would also make the features field exported (unfortunately this means we need a different name for it). Then Add is not needed.

mnsmar replied:
ok but we still need a way to access the transcripts right? If I understood correctly previously you suggested to populate the (public) `Transcripts` field with `func gene.Transcripts(feat.Set) []gene.Transcript`. The question is when is this function called if not from `Add`? Is it the clients responsibility? cause if it is, don't you think its kind of cumbersome?

kortschak replied:
All you need is `func TranscriptsOf(feat.Set) []Transcript` that can be called by the client. I don't think it is cumbersome.

Consider one use case in client code where `g` is a gene.Gener (I'm thinking this name might need to change - gene.Interface?):

```go
for _, t := range gene.TranscriptsOf(g) {
    analyseTranscript(t)
}
```

mnsmar replied:
ok I get. So, correct me if I'm wrong, you are trying to keep the `Gene` API lean and offer a more generic function. I have to get used thinking this way instead of pure OO. 
So `Gener` -> `Interface` and `Transcripter` -> `Transcript` or just the former?

kortschak replied:
I would do both.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,125 @@
+}
+
+// Add adds a feature to the gene. Currently it supports transcripts only.
+func (s *Gene) Add(f ...feat.Feature) error {
```
Ditto for if this is made public.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,125 @@
+
+// AddTranscript will add a transcript to the gene.
+// Will return an error if it encounters one.
+func (s *Gene) AddTranscript(t Transcripter) error {
```
I think the Gene should know its orientation before adding a transcript and that it should be reasonable to add antisense transcripts (the semantics of what constitutes a gene should not be our business).

The new feat.OrientationOf function (biogo/biogo@8fbc4dc3ab530a61295995bd6d0708ee3b480239) can be used to get orientation of transcripts and other elements associated with a Gene.

mnsmar replied:
If biologists only knew what developers imply in the code! To take it a step further - is Gene even a feat.Orienter? At least for now if a gene's orientation does not match that of its transcripts I don't see the reason why it should have an orientation in the first place.

kortschak replied:
Mmm.

I think Gene should be a feat.Orienter. Then all added Transcripts should (depending on your definition of a gene) just return the feat.Forward from feat.OrientationWithin(transcript, transcript.Location()). Though some might argue that antisense transcripts might be considered part of a gene.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,125 @@
+		return errors.New("transcript out of gene bounds")
+	}
+
+	if s.Location() == nil {
```
The transcript should not care about this.

kortschak commented on feat/gene/gene_test.go:
```diff
@@ -0,0 +1,203 @@
+)
+
+//////////////////////////////////////////////////////////////////////////////
+//////////////////////////////////////////////////////////////////////////////
```
Please remove comment decoration.

kortschak commented on feat/gene/gene_test.go:
```diff
@@ -0,0 +1,203 @@
+	for _, d := range geneTestData {
+		g := d.Gene
+
+		if g.Name() != d.Name {
```
Revise these to use gopkg.in/check.v1.

kortschak commented on feat/gene/transcript.go:
```diff
@@ -0,0 +1,498 @@
+// Copyright ©2012 The bíogo Authors. All rights reserved.
```
Date.

kortschak commented on feat/gene/transcript.go:
```diff
@@ -0,0 +1,498 @@
+//////////////////////////////////////////////////////////////////////////////
+// A NonCodingTranscript is a gene transcript that does not have coding
+// potential.
+type NonCodingTranscript struct {
```
s/NonCodingTranscript/NonCoding/g

mnsmar replied:
I do not like the long name either but in this case don't you think that it makes things cleaner. From the clients perspective I think gene.NonCodingTranscript  is better than gene.NonCoding. The latter gives the impression that it refers to a gene, I would definitely be puzzled reading it.

kortschak replied:
OK

kortschak commented on feat/gene/transcript.go:
```diff
@@ -0,0 +1,498 @@
+}
+
+//////////////////////////////////////////////////////////////////////////////
+type Utr5 struct {
```
s/Utr5/UTR5/g

kortschak commented on feat/gene/transcript.go:
```diff
@@ -0,0 +1,498 @@
+	return nil
+}
+
+//////////////////////////////////////////////////////////////////////////////
```
s/Cds/CDS/g

kortschak commented on feat/gene/transcript.go:
```diff
@@ -0,0 +1,498 @@
+}
+
+//////////////////////////////////////////////////////////////////////////////
+type Utr3 struct {
```
s/Utr3/UTR3/g

kortschak commented on feat/gene/transcript.go:
```diff
@@ -0,0 +1,498 @@
+//////////////////////////////////////////////////////////////////////////////
+// A CodingTranscript is a gene transcript that has coding potential.
+// It defines 5'UTR, CDS and 3'UTR
+type CodingTranscript struct {
```
s/CodingTranscript/Coding/g

kortschak commented on feat/gene/transcript.go:
```diff
@@ -0,0 +1,498 @@
+	ChromOrientation feat.Orientation
+	exons            []*Exon
+	introns          []*Intron
+	utr5             *Utr5
```
It seems to me that utr5, cds and utr three can be described by two ints (or three []*Exon). Dropping the Utr5, Cds and Utr3 types completely.

mnsmar replied:
You are right but if for example a client requests the coding sequence from the transcript (which is highly likely) then the transcript would have to build it from the information it has (the 2 ints) which is bad because that would be an implementation dependency. I think that we could squash the 3 types into a single one that satisfies SpliceableFeature though. What do you think?

kortschak replied:
If we think about the most complex gene we can. We undoubtedly have an entity with less than 1000 exons. Even just iterating over the slice of features describing exons and introns to find those that are within the interval will be cheap and only depends on the feat.Range interface.

mnsmar replied:
Definitely, but my note was not so much about performance but responsibility. Is CDS responsible to answer questions about itself like somecds.overlaps(somefeature) or is it the transcript's responsibility sometran.does_cds_overlap(somefeature)?

kortschak replied:
I think it's either the responsibility of the package or the Gene.

mnsmar replied:
I lost you here. The Gene has nothing to do with it - it does not know anything about coding sequence. Only CodingTranscript knows and should know about it. I will need to come up with a user case to clarify what I'm thinking but it's getting really late and my brain probably malfunctions. I'll catch up later.

kortschak replied:
Yes, sorry. I mean CodingTranscript. I too am fried.

kortschak commented on feat/gene/transcript.go:
```diff
@@ -0,0 +1,498 @@
+//////////////////////////////////////////////////////////////////////////////
+// An Intron is the part of a transcript that is removed during splicing
+// and is not part of the final mature RNA product.
+type Intron struct {
```
Intron methods on concrete rather than pointer type.

kortschak commented on feat/gene/transcript.go:
```diff
@@ -0,0 +1,498 @@
+//////////////////////////////////////////////////////////////////////////////
+// An Exon is the part of a transcript that remains present in the final
+// mature RNA product after splicing.
+type Exon struct {
```
Exon methods on concrete rather than pointer type.

mnsmar replied:
That's interesting. Why is that? Any particular benefit of one over the other?

kortschak replied:
You don't have any mutating methods on Exon and it's relatively small.

kortschak commented on feat/gene/transcript.go:
```diff
@@ -0,0 +1,498 @@
+}
+
+type Splicer interface {
+	Exons() []*Exon
```
Exons() []Exon
Introns() []Intron

kortschak commented on feat/gene/transcript.go:
```diff
@@ -0,0 +1,498 @@
+	Introns() []*Intron
+	ExonicLen() int
+	IntronicLen() int
+	AddExon(...*Exon) error
```
AddExon(...Exon)

kortschak commented:
I've done a first pass, but I need to go over it again when comments have been addressed.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,125 @@
+type Gener interface {
+	feat.Feature
+	feat.Orienter
+	Features() []feat.Feature
```
s/Features() []feat.Feature/feat.Set/

mnsmar commented on feat/gene/transcript.go:
```diff
@@ -0,0 +1,498 @@
+// potential.
+type NonCodingTranscript struct {
+	Id               string
+	Chrom            feat.Feature
```
For the types `Gene`, `CodingTranscript` and `NonCodingTranscript` I used Chrom as a prefix for the location coords. Do you think I should use Loc as I did for `Intron` and `Exon`?

kortschak replied:
I think this is fine.

mnsmar commented:
I will also put everything in a single file `gene.go`. Is that ok?

kortschak commented:
SGTM

mnsmar commented:
Addressed all the comments (hope so) + some extras. I need to add a few tests for the CDS and UTRs of `CodingTranscript` but I will probably do it tomorrow.

mnsmar commented:
Done with all changes.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,365 @@
+// as 5'UTR, CDS and 3'UTR. Transcripts also consist of exons and introns. The
+// former are elemnts that are part of the final mature RNA product whereas
+// the latter are not.
+
```
Delete line - you will not see the package doc comment.

Also, I think we can lose some of the philosophy and just say what we define a gene as.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,365 @@
+}
+
+// OrientedFeature defines a feat.Feature that also has orientation.
+type OrientedFeature interface {
```
s/OrientedFeature/FeatureOrienter/g

kortschak replied:
Actually, this is only used in tests. Delete it from here and have it unexported in the tests.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,365 @@
+}
+
+// SpliceableFeature defines a feat.Feature that undergoes splicing
+type SpliceableFeature interface {
```
This is not used anywhere except in Transcript and the contracts. Delete it.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,365 @@
+// Transcript is the interface for a gene transcript.
+type Transcript interface {
+	feat.Orienter
+	SpliceableFeature
```
Replace feat.Orienter with feat.FeatureOrienter and SpliceableFeature with Splicer.

mnsmar replied:
Can't find `feat.FeatureOrienter`. You want me to create it? If yes, I can also get rid of the one I define in this package.

kortschak replied:
See [comment at line 39](https://github.com/biogo/biogo/pull/1#discussion-diff-23576385R39), but in light of that, just explicitly use `{ feat.Orienter;feat.Feature; Exons() Exons; Introns() []Intron }`.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,365 @@
+
+// IntronicLen returns the length of the intronic region of the transcript.
+func (s *CodingTranscript) IntronicLen() int {
+	exoniclen := s.ExonicLen()
```
We don't need two lines here.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,365 @@
+	return i.LocOrientation
+}
+
+//////////////////////////////////////////////////////////////////////////////
```
Remove decoration.

kortschak commented on feat/gene/gene_test.go:
```diff
@@ -0,0 +1,362 @@
+// Copyright ©2014 The bíogo Authors. All rights reserved.
```
Fix date.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,365 @@
+// Copyright ©2014 The bíogo Authors. All rights reserved.
```
Fix date.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,365 @@
+}
+
+// ExonicLen returns the length of the exonic region of the transcript.
+func (s *NonCodingTranscript) ExonicLen() int {
```
I'm thinking perhaps rather make a type Exons []Exon and give it a SplicedLen method that does this.

mnsmar replied:
Should the `Exons() []Exon` method change to `Exons() Exons`?

kortschak replied:
Yes, that's needed to allow a g.Exons().SplicedLen().

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,365 @@
+}
+
+// IntronicLen returns the length of the intronic region of the transcript.
+func (s *NonCodingTranscript) IntronicLen() int {
```
Is this biologically meaningful?

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,365 @@
+}
+
+// ExonicLen returns the length of the exonic region of the transcript.
+func (s *CodingTranscript) ExonicLen() int {
```
See comment above for NonCodingTranscript ExonicLen.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,365 @@
+
+//////////////////////////////////////////////////////////////////////////////
+// ExonFits checks if an exon is within feature bounds
+func ExonFits(ref feat.Feature, exons ...Exon) error {
```
This should return a bool.

mnsmar commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,365 @@
+}
+
+// Splicer defines the interface for elements that undergo splicing.
+type Splicer interface {
```
I could also let this go. It is only used in `Transcript`.

kortschak replied:
Sounds good to me.

mnsmar commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,365 @@
+
+// An Intron is the part of a transcript that is removed during splicing
+// and is not part of the final mature RNA product.
+type Intron struct {
```
There is currently no connection between `Transcript` and `Intron` (no `AddExon` equivalent). Do you think I should make an `AddIntron` or should the `Transcript` create them automatically from the exon information? The latter is against the Dependency Injection Principle. Of course there is the option to drop `Intron` altogether.

kortschak replied:
I don't think in OO, so in general the word "Principle" will invariably leave me feeling sceptical. It seems to me, that if there is an efficient way to calculate the introns from the exons (and given the space complexity of exonic structure that's a reasonably assumption) then we should just do that. This means that the []Introns will always be correct for a given []Exon at the time of the call. If people want to amortise the (minuscule) costs, they can cache the result themselves.

mnsmar replied:
Sorry it's a design pattern, not principle, and applies not only to OO. I helps create decoupled systems. In this case I will agree with you though that it makes sense to let the transcript create the introns from the exons. Do you think that I should also make a `type Introns []Intron` for consistency?

kortschak replied:
I understand it's a pattern. "Patterns" as exist in OO are largely eschewed in Go. There are patterns that are beginning to be noted, but generally, the main pattern is "Does this sanely describe what I am trying to model?"

If there are no methods that are meaningful on a []Intron then it doesn't need a type declaration. Though if there might be one in the future that has biological meaning then maybe.

I was thinking more about the dynamic calculation and it seems to me that a transcript is really best defined as a collection of spliced exons. This means that it's feature is best defined dynamically by calculating along its []Exon. This together with the dynamic calculation of introns means that exons should be stored in sort order (at least during the calculation. They should also be non-overlapping (this is an invariant rather than an implementation detail).

mnsmar replied:
I am not aware of a biologically relevant method on []Intron either. But mainly for consistency and to a lesser extent support possible relevant methods in the future I think it makes sense to do a type declaration at this stage. I personally would prefer knowing that Exons() and Introns() return similar things. Agree?

I was waiting to clean up the comments on this round to describe exactly this, that the exons should be sorted and non overlapping. ok let me clean up the code to discuss about the dynamic nature of the transcript on a clean base.

kortschak replied:
>I personally would prefer knowing that Exons() and Introns() return similar things. Agree?

OK

mnsmar commented:
> I was thinking more about the dynamic calculation and it seems to me that a transcript is really best  defined as a collection of spliced exons. This means that it's feature is best defined dynamically by calculating along its []Exon. This together with the dynamic calculation of introns means that exons should be stored in sort order (at least during the calculation. They should also be non-overlapping (this is an invariant rather than an implementation detail).

So how do you think a transcript's feature should be calculated from exons? This also implies that the location of an Exon cannot be the transcript. Should we drop `ChromStart`, `ChromEnd`, from transcripts?

mnsmar commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,330 @@
+
+// Append offers fine grained control on how appending works.
+// Should check for sorting and overlap.
+func (s *Exons) Append(exons ...Exon) error {
```
Created this here only as a placeholder. How do you think the sorting and overlap restrictions should be applied on `Exons`.

kortschak replied:
Need to think about that.

kortschak replied:
I guess it means we need to keep exons unexported and make checks during addition. I think `Add` is better than `Append` there since it may be done out of order.

append, sort then check overlap for each Add call - note in docs that this is done so people know that it's cheaper to add en bloc.

kortschak commented:
>So how do you think a transcript's feature should be calculated from exons?

start = min(exon starts) where exon.loc == transcript
end = max(exon ends) where exon.loc == transcript

There is no point sorting prior to this obviously. But a call to sort.IsSorted() and a conditional sort.Sort() prior to determining introns is needed. This means we also need (probably not exported - say `exons`) a type that satisfies sort.Interface.

>This also implies that the location of an Exon cannot be the transcript. Should we drop ChromStart, ChromEnd, from transcripts?

Yes.

mnsmar commented:
> start = min(exon starts) where exon.loc == transcript
I'm confused by this. If exon.loc is the transcript then how do i get the transcript.start from it?
I was thinking something like the following:

    if exon.loc == transcript {
        error
    }
    transcript.start = min(exon starts)
    transcript.loc = exon.loc

kortschak commented:
e.g.

```go
func (t *CodingTranscript) Start() int {
    s := maxInt
    for _, e := range t.exons {
        if e.loc != t {
            continue
        }
        if e.Start() < s {
            s = e.Start()
        }
    }
    return s
}
```

Not sure whether to panic if there is no exon.

mnsmar commented:
1. Shouldn't t.Location() == exon.Location()? If that's the case then the second `if` clause results in a circular reference. We return the transcript start on the transcript itself.
2. Also the first if clause prevents legit cases where the exons location is not the transcript (eg. chromosome). I would expect `transcript.Start()` be the min(exon starts) and `t.Location==chromosome`. 
Like so:
``` go
func (t *CodingTranscript) Start() int {
    s := maxInt
    for _, e := range t.exons {
        if e.loc == t { // notice I changed not equal to equal
            // Maybe panic cause this should not happen or at least set the t.Location to nil
        }
        if e.Start() < s {
            s = e.Start()
        }
    }
    return s
}
```

kortschak commented:
You could do it that way, but i like to think of Location as not just a physical location, but also a conceptual location. The exon is located within a transcript, which is located in a gene, which is located in a chromosome.

mnsmar commented:
I agree with you and the implementation we have so far allows for a tree-like structure or any other structure. It's up to the client, how to build the structure. However I'm now thinking that the dynamic nature of the transcript might not be such a good idea after all. Below I give some statements/test cases and you can tell me if we are on the same page.

1. I expect that the coords (`Start`, `End`) of a Feature are defined in respect to `Location`. This means that when I write `Feature {Start : 10, Location: AnotherFeature}` it means that `Feature` starts 10 units away from the start of `AnotherFeature`. This also means that function `BaseStartOf()` has a specific meaning and can be defined.
2. The dynamic location of a transcript must match that of its exons. Otherwise strange things could happen as in the following snippet:

 ``` go
 // Build tree-like structure
 chromA = Chrom{
     Start: 0
     Location: nil
 }
 transcA = Transcript{
     Location: chromA    // we allow sth different than exonA.Location
 }
 exonA = Exon{
     Start:200
     Location: transcA
 }
 print(transcA.Location) // prints chromA 
 print(transcA.Start)    // prints 200. BOOM!
 ```

3. Given (2) the tree-like capabilities for transcript break. The transcript can no longer be on anything other than the exon location.

 ``` go
 chromA = Chrom{
     Start: 0
     Location: nil
 }
 transcA =Transcript{
 }
 ExonA = {
     Start:200
     Location: transcA
 }
 print(transcA.Location) // prints transcA
 print(transcA.Start)    // prints 200. Does this really make sense?
                         // What about (1)? Exon was expected to be 200 units away from
                         // the transcript start but this is no longer the case.
 ```

kortschak commented:
>It's up to the client, how to build the structure.

That's not entirely true. Sure, it's up to them, but there are certain semantic expectations that are (as yet poorly documented) the basis for the design. It's reasonable to break these internally (I do in some packages), but not to expose them publically.

1. I agree that locations relative to the location is a reasonable expectation (that's not defined - perhaps add it to your doc comments) and if that's what we want then the dynamic calculation is not the way to go and we need to hide much of the currently exposed fields. Alternatively, if the trasncript types keep an offset to translate the coordinates between chromosomal and transcript that would work - this is how I do coordiate mapping for the sequence types. Perhaps try that first.

2. An offset fixes that: just return the previous func's return value plus the offset.

3. ditto with the caveat that that is clearly broken; transcA has no location.

mnsmar commented:
Can you please give an some examples of semantics that are poorly annotated and that should be restrictions to the client. Just so I can get a better understanding on what you refer to. 

Maybe I'm missing some of the biogo internals but I cannot see any reason to introduce an offset. I'm mostly happy with the current implementation. Things are decoupled. Every type can stand on each own (eg a transcript works without necessarily needing the exons - although maybe it will panic if we call a function that needs them). The coords/location structure is flexible. To me right now an offset would just add (mental) complexity without an evident advantage. What is that advantage?

I did look at `seq.linear.Seq` and some others and realized that you use `OffSet` the same way we use `ChromStart`. Is that right? The word "global" that you use in the docs puzzles me. Does it refer to the sequence's location or the "BaseLocation"? I will try to dive in the sequence types some more to try to understand how exactly you use OffSet. Meanwhile if you have some insight that could help me that would be great.

kortschak commented:
The semantics that I'm refering to are that levels the location chain should not be skipped, i.e. transcript should be between chrom and exon (possibly keeping gene in between transcript and chrom if that is biologically meaningful to the problem).

The advantage is that I would like to keep the hierarchical relationship between exon and transcript. I'm not sure that exon and transcript are really separable in a conceptual sense; a transcript must have at lease one exon (even a bacterial transcript). Essentially a transcript is an annotation on a number of exons. You give a couple of cases where positions don't work properly if the location chain works the way I want, a single offset stops that happening.

Yes, offset is the ChromStart. You are right, that wording is terrible; it refers to the location. This should probably be fixed, and adding a generalised offset-relative-to-containing-feature similar to OrientationWithin would be helpful. I'm happy to be swayed on ways this should work.

kortschak commented:
To add to the above, there is a similar issue with the relationship between transcripts and genes.

Take this example: you construct a gene/transcripts/exons that is located on one chromosome, but you want to lift it over to another location because of some structural variation. If the locations are all pinned onto the gene (rather than directly to the chromosome), the only values that need to be altered are the offset and location of the gene. Otherwise, you need to update all coordinates and locations or have non-sane reutrns for Start()/End()/Location().

mnsmar commented:
I'm writing this really fast just as food for though (and so as not to forget) cause I'm fried and need to get some rest. What if we rename `CromStart` and the likes in all our types to `Pos` define an `OffSet` for all types and then "define" that the `Start` is always `OffSet + Pos` (`Pos` will be calculated dynamically for transcripts and will be set at creation for the rest). I haven't though of it thoroughly but I think it gives consistency. I don't like having an OffSet value just for `Transcript` and in a way "surprise" the client on what this actually means. I repeat - I haven't though of it thoroughly :)

Regarding your latest comment: I see what you mean, I will get back to you later but just a quick comment. On the other hand if the client chooses to define everything on the chromosome think of the performance boost. Does not have to traverse deep in the tree to get the coords. This can be a important in the high throughput sequencing world that we now live.

kortschak commented:
Thanks for working through this with me, I appreciate your patience. I agree with everything you have there. I was also considering the performance impact, but I think I fall on the side that I'm prepared to accept that from an out of the box generalised implementation. When people want ultimate performance, they should really be implementing far more specific types themselves (types are cheap in Go).

mnsmar commented:
That's fine, I enjoy diving in the semantics and I understand that the initial design is perhaps the single most important thing to create sth that scales well. Personal story: A couple years back I tried to explain to a guy why a framework I was working on would be useful. So I explained the gene structure, transcripts, splicing, etc only to get the reply "Oh come on, you just created some gene models". Seriously? Maybe I should forward this PR to the guy!

So let's summarize what needs to be done so I can get on it. These are my suggestions.

1. Rename all fields with suffix "Start" to `Pos`.
2. Delete all fields with suffix "End" and introduce `Length` instead.
3. Introduce `OffSet` for all types.
4. Make clear in the docs and implement that 'Start == Pos + OffSet` and that it refers to `Location()`
5. Couple `Exon` and `Intron` with `Transcript`. Maybe rename `Loc` field to `Transcript` to make the coupling more evident.
6. Keep `Transcript` decoupled from `Gene` - Rename transcript field `Chrom` to `Loc`. I think leaving the coupling of these two to the client is important. A transcript can definitely be defined without a gene and vice versa. As a matter of fact the UCSC GTF file that supposedly contains the genes contains only transcripts.
7. Sort Exons and ensure they are not overlapping.

mnsmar commented:
oh my, the more I try to justify the OffSet to myself the more I think it should not be used. I even tried to explain it to a colleague of mine and he was muddled. The notion that I always end up to, is that a transcript has a position onto itself (this is basically what the dynamic `Pos` boils down to) which is just mind blowing to me. For some reason my brain cannot accept this idea.

Think about this: We enforce that the exons added to a transcript must much the transcript itself. Also the `AddExon` method becomes `AddExons` and can only be used if the `exons` field is empty. So we enforce that all exons are added at the same time. And here comes the important thing - there must be one exon that starts at 0.

```go
t = Transcript{
    Loc : a chromosome or a gene
    Pos: 1000
}
t.AddExons(Exon{Loc: t, Pos: 0, Len: 100}, Exon{Loc: t, Pos: 500, Len:1500})

print(t.Loc) // chromosome or gene
print(t.Start) // 1000
print(t.End) // 2000
```

Tight coupling without the need for dynamic calculation or offset. What do you say?

kortschak commented:
>The notion that I always end up to, is that a transcript has a non zero position onto itself (this is basically what the dynamic Pos boils down to) which is just mind blowing to me.

This is not correct. It has a non-zero position relative to its location.

>Tight coupling without the need for dynamic calculation or offset. What do you say?

SGTM but call it SetExons and allow it to be called again, blowing away the previous collection.

What do you intend to do with the equivalent relationship between transcripts and genes?

I'm confused about why you think there is no need for an offset - that's what Pos is. The reason I didn't use pos in seq is that pos (despite its use in SAM doesn't match a non-unit lengthed object well IMO), hence offset.

mnsmar commented:
> SGTM but call it SetExons and allow it to be called again, blowing away the previous collection.

better

> What do you intend to do with the equivalent relationship between transcripts and genes?

If we agree that the gene cannot exist on its own I could do the same thing using its Features. Sounds good?

> I'm confused about why you think there is no need for an offset - that's what Pos is. The reason I didn't use pos in seq is that pos (despite its use in SAM doesn't match a non-unit lengthed object well IMO), hence offset.

Yes sorry in a previous comment I mentioned that `Start == Pos + OffSet`. Now we got rid of one of these guys, that's what I really wanted to say. You prefer to name it `OffSet` then right?

kortschak commented:
>If we agree that the gene cannot exist on its own I could do the same thing using its Features. Sounds good?

SGTM

>Yes sorry in a previous comment I mentioned that Start == Pos + OffSet. Now we got rid of one of these guys, that's what I really wanted to say. You prefer to name it OffSet then right?

Yes please, but lc/S/.

mnsmar commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,426 @@
+// A Gene occupies a specific region on the genome and may have 0 or more
+// features, including transcripts, attached to it. The gene coordinates are
+// independent of the features it contains.
+type Gene struct {
```
I kept the Gene decoupled from its Features. It's coordinates must be specified at creation. I think for the Gene it makes more sense this way.

kortschak replied:
Why? The gene is functionally defined by its features. If the client wants to make a broad-brush definition of the gene, they can add an ecompassing feature - in Gene the features should be able to overlap, so this is not an issue.

mnsmar replied:
ok.
I don't know how to keep the structure consistent when the client attempts to change one of the features?
e.g.
```go
trA := &NonCodingTranscript{Loc: geneA, Offset: 0}
trB := &NonCodingTranscript{Loc: geneA, Offset: 10}
g.SetFeatures([]feat.Feature{trA, trB})
// At this point the gene is consistent because SetFeatures does all the required checks

trA.Loc = Chrom("X")
// now things are getting hairy
```
1. One option would be to create non exposed fields `offset`, `location`, ... in `Gene` that would be set when `SetFeatures` is called and never change after that (or maybe change when SetFeatures is called again). This way when one transcript changes the gene is not affected.
2. Another option would be to do a consistency check every time `Start()` or `Location()` etc are called.
3. Or we could hide all fields and make everything immutable.

Any ideas?

kortschak replied:
There is only so much we can do. The principle at play is that we should make it easy for the client to do the right thing (not require that they keep additional book-keeping or do additional checks), but not that we make it impossible to do the wrong thing. When we synthesis data from the stored collections the situation above should probably result in a panic (this needs some thought though).

mnsmar replied:
I agree with you up to a point but I think that since we know that such a situation might occur we should take some measures to avoid it. I tend to think that we should promote immutability for the location related attributes (`Location`, `Start`, `End`, `Orientation`). Once set either explicitly or implicitly through `SetExons` or `SetFeatures` they should not be allowed to change. Of course this means hiding some fields but I do not see any major drawback enforcing such restrictions. The client that wants to change these attributes should probably clone to a new variable. Also from my experience I usually see that the gene structure is set at the beginning of a script and never needs to be tampered with.

kortschak replied:
Make them unexported hampers serialisation.

If you are performing operations on a transcripomic data set (read mappings), you may be constructing gene models). Having to continually reclone is onerous and will cause excessive allocations, the alternative is to keep details in another structure until you are ready - defeating the purpose of this set of types.

mnsmar replied:
I don't think that these types are suitable in the case you mentioned either way but maybe I'm wrong. I like to see of these types as more rigid building blocks for modelling genomic annotation (coming from any source e.g. UCSC, Ensembl, custom, read mappings) and then using them to ask questions such as "how many reads are contained in the exonic area covered by gene X".  Anyway this is opinion-based and I don't think there is a very high chance a client is gonna blow the structure so I'm fine with it. Please review the code and let me know if anything needs to change.

mnsmar commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,426 @@
+}
+
+// UTR5start returns the start of the 5'UTR relative to the transcript.
+func (t *CodingTranscript) UTR5start() int {
```
I'm not sure if that's the best way to go. Maybe we should re-introduce utr5, cds, utr3 types (maybe hidden) and let transcript create them dynamically from `CDSstart` and `CDSend` and return them as feat.feature. For eg. `func (t *CodingTranscript) UTR3() feat.Feature`. The way we have it now, we delegate everything to transcript and I don't think this scales very well.

kortschak replied:
If you do that, then either the type or the client needs to now coordinate a number of elements that must not overlap by the invariants implicit in a gene definition. It seems to me that this does that already and is either cheap because the n of the elements is small, or can be made cheap if that becomes an issue (I don't believe it will).

mnsmar replied:
OK let's leave it like this for now and see how it evolves.

kortschak replied:
SGTM

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,457 @@
+// Location is the gene and that one of them has zero Start. If an error
+// occurs it is returned and the features are not set.
+func (g *Gene) SetFeatures(feats ...feat.Feature) error {
+	pos := 1000000000
```
Make this maxInt and minInt. (Negative values have no special meaning in biogo, so we allow them).

kortschak replied:
Oh, sorry, just pos := maxInt.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,457 @@
+	} else if t.Orient == feat.Reverse {
+		return t.CDSend
+	}
+	return 0
```
This should probably be a panic. A zero orientation is not reasonable for a coding transcript.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,457 @@
+	} else if t.Orient == feat.Reverse {
+		return t.Len()
+	}
+	return 0
```
Same as above.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,457 @@
+	} else if t.Orient == feat.Reverse {
+		return 0
+	}
+	return 0
```
Same as above.

kortschak commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,457 @@
+	} else if t.Orient == feat.Reverse {
+		return t.CDSstart
+	}
+	return 0
```
Same as above.

kortschak commented:
OK, I've had a brief review. I'll look again in a couple of days, but it's looking pretty good.

mnsmar commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,458 @@
+// Location is the gene and that one of them has zero Start. If an error
+// occurs it is returned and the features are not set.
+func (g *Gene) SetFeatures(feats ...feat.Feature) error {
+	pos := util.MaxInt
```
Is this what you mean?

kortschak replied:
Yeah, I'd just use the actual expression to avoid the import - maybe just copy the definition from util as an unexported const.

mnsmar commented:
Addressed all comments. Waiting for next review and directions on how to proceed with the merge (rebasing, etc).

mnsmar commented on feat/gene/gene.go:
```diff
@@ -0,0 +1,459 @@
+// latter is guarranted by setting the gene end at the largest end of the
+// features.
+type Gene struct {
+	Id     string
```
I should change `Id` to `ID` for all types right?

kortschak replied:
Yeah, probably. I have that wrong in places.

kortschak commented:
How is this coming along?

mnsmar commented:
I've finished all changes. It awaits your review and let me know how to proceed in terms of rebasing.

kortschak commented:
LGTM

Just rebase onto master squashing into one commit. When that's done I'll do a ff merge.

mnsmar commented:
Done

kortschak commented:
Thanks.

mnsmar commented:
Welcome. Thanks for the guidance
